source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -340,7 +340,7 @@ const std::string& SslSocket::urlEncodedPemEncodedPeerCertificateChain() const {
     return cached_url_encoded_pem_encoded_peer_cert_chain_;
   }
 
-  STACK_OF(X509)* cert_chain = SSL_get_peer_full_cert_chain(ssl_.get());
+  STACK_OF(X509)* cert_chain = SSL_get_peer_cert_chain(ssl_.get());
```

Comment:
If I understand correctly, this changes the behavior of `urlEncodedPemEncodedPeerCertificateChain` such that it will only return the intermediate certificates, and not the leaf certificate, when called from the server context.

The documentation added for the new field ([here](https://github.com/envoyproxy/envoy/pull/7026/files#diff-db458208464b55cb68c1f038faf7dc23R136) and [here](https://github.com/envoyproxy/envoy/pull/7026/files#diff-62f92f4fe8350a66ff9e3ca9e5299228R331)) suggests that it will be the "entire client cert chain", which could be easily interpreted as including the leaf cert.

The current documentation for `urlEncodedPemEncodedPeerCertificateChain` suggests that it should return the leaf certificate as well:

https://github.com/envoyproxy/envoy/blob/cbd8d57e9fc2921d26236db24b6bf91246ba3b4f/include/envoy/ssl/connection.h#L68-L72

From what I could tell nothing outside of tests is currently using this method, so a behavior change is probably okay, but the docs would need to be rationalized.

On the other hand, would it make sense to retain the previous behavior? Is the motivation to avoid duplicating information in case somebody requests both the `cert` and `chain` fields for the xfcc header? My intuition would be that if somebody wants the full chain, they would set `chain: true` and `cert: false` to get the full chain in one field.

Reply:
It seems that OpenSSL doesn't have a `SSL_get_peer_full_cert_chain()` API while BoringSSL does. This change is made in order to prevent breaking the code when compiling with these libraries.

Reply:
Also, thanks for catching the doc errors. I've addressed those in 64306bb.

Reply:
Thanks @venilnoronha. If we use the openssl behavior, there's some nuance in the documentation for [`urlEncodedPemEncodedPeerCertificateChain`](https://github.com/envoyproxy/envoy/pull/7026/files#diff-55efddead9c2a3a1f5191d1f310a67b0R70) in that it will return the full peer cert chain for the client half of a connection but will omit the leaf cert for the server half.

See https://github.com/google/boringssl/blob/1b878e7cc69f60407457024a212750c52bcc9ae5/ssl/ssl_x509.cc#L537-L550 and https://github.com/openssl/openssl/blob/c04b66b18d1a90f0c6326858e4b8367be5444582/ssl/ssl_lib.c#L1545-L1560.

That nuance seems like complexity that callers of this API shouldn't need to understand. My understanding is that the OpenSSL build is being implemented as an extension, and that it could be written to create the full cert chain by pushing the leaf cert onto the stack of certs returned by `SSL_get_peer_cert_chain` if operating as a server to get the same behavior.

Reply:
Thanks for highlighting that. I've reverted the change for now.